File: generated_code/default/generalization/Java/Adapter.java
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.apache.http.HttpResponse;
import org.apache.http.NameValuePair;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.concurrent.FutureCallback;
import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
import org.apache.http.impl.nio.client.HttpAsyncClients;
import org.apache.http.message.BasicNameValuePair;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;


public class Adapter {
    private static String serverIP = "0.0.0.0";
    private static String port = "2000";
    //static String serverUrl = "https://httpbin.org/post";
    private static String serverUrl = "http://" + serverIP + ":" + port + "/";
    //private static CloseableHttpClient httpClient = HttpClients.createDefault();
    private static CloseableHttpAsyncClient httpAsyncClient = HttpAsyncClients.createDefault();

    static {
        httpAsyncClient.start();
    }

private static List<NameValuePair> basicDataArgs(JsonObject data) {
        if(data == null)
            throw new NullPointerException("data cannot be Null");
        String dataStr = data.toString();
        List<NameValuePair> dataArgs = new ArrayList<NameValuePair>();
        if(data.has("_id")) {
            dataArgs.add(new BasicNameValuePair("_id", data.get("_id").getAsString()));
        }
        else {
            dataArgs.add(new BasicNameValuePair("data", dataStr));
        }
        return dataArgs;
    }

    private static List<NameValuePair> basicDataArgs(JsonArray data) {
        if(data == null)
            throw new NullPointerException();
        String dataStr = data.toString();
        List<NameValuePair> dataArgs = new ArrayList<NameValuePair>();
        dataArgs.add(new BasicNameValuePair("data", dataStr));
        return dataArgs;
    }

    public static void readOne(String collection, JsonObject data, Object CBModel) {
        List<NameValuePair> dataArgs = basicDataArgs(data);
        callServer(collection, "readOne", dataArgs, CBModel);
    }

    public static void readAll(String collection, JsonObject data, Object CBModel) {
        List<NameValuePair> dataArgs = basicDataArgs(data);
        callServer(collection, "readAll", dataArgs, CBModel);
    }

    public static void create(String collection, JsonObject data, Object CBModel){
        List<NameValuePair> dataArgs = basicDataArgs(data);
        callServer(collection, "create", dataArgs, CBModel);
    }

    public static void create(String collection, JsonArray data, Object CBModel){
        List<NameValuePair> dataArgs = basicDataArgs(data);
        callServer(collection, "create", dataArgs, CBModel);
    }

    public static void update(String collection, JsonObject data, Object CBModel){
        if (!data.has("newData")) {
            throw new IllegalArgumentException("lack key 'newData' in data");
        }
        List<NameValuePair> params = new ArrayList<NameValuePair>();
        if(data.has("_id")) {
            params.add(new BasicNameValuePair("_id", data.get("_id").getAsString()));
        }
        else if(data.has("oldData")){
            params.add(new BasicNameValuePair("oldData", data.get("oldData").toString()));
        }
        else {
            throw new IllegalArgumentException("lack key '_id' or 'oldData' in data");
        }
        params.add(new BasicNameValuePair("newData", data.get("newData").toString()));
        callServer(collection, "update", params, CBModel);
    }

    public static void delete(String collection, JsonObject data, Object CBModel){
        List<NameValuePair> dataArgs = basicDataArgs(data);
        callServer(collection, "delete", dataArgs, CBModel);
    }

    public static void callServer(String collection, String operation, List<NameValuePair> params, Object CBModel){
        String url = serverUrl + operation;
        params.add(new BasicNameValuePair("collection", collection));
        PostCall(url, params, CBModel);
    }

    //java.io.UnsupportedEncodingException
    public static void PostCall(String url, List<NameValuePair> params, Object CBModel){
        Method successM, errorM;
        try {
            successM = CBModel.getClass().getMethod("successCB", String.class);
            errorM = CBModel.getClass().getMethod("errorCB", String.class);
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
            return;
        }

        HttpPost post = new HttpPost(url);
        post.setHeader("Content-Type", "application/x-www-form-urlencoded; charset=utf-8");

        if (params != null) {
//            StringEntity params = null;
//            try {
//                params = new StringEntity(data.toString());
//            } catch (UnsupportedEncodingException e) {
//                e.printStackTrace();
//            }
//            post.setEntity(params);
            try {
                post.setEntity(new UrlEncodedFormEntity(params,"utf-8"));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }
        }
        httpAsyncClient.execute(post, new FutureCallback<HttpResponse>() {
            @Override
            public void completed(HttpResponse result) {
                try {
                    String s = new BufferedReader(new InputStreamReader(result.getEntity().getContent()))
                            .lines().collect(Collectors.joining(System.lineSeparator()));
                    successM.invoke(CBModel, s);
                } catch (IllegalAccessException | InvocationTargetException | IOException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void failed(Exception ex) {
                try {
                    errorM.invoke(CBModel, ex.getMessage());
                } catch (IllegalAccessException | InvocationTargetException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void cancelled() {
                try {
                    errorM.invoke(CBModel, "cancelled");
                } catch (IllegalAccessException | InvocationTargetException e) {
                    e.printStackTrace();
                }
            }
        });

    }


}

File: generated_code/default/generalization/JavaScript/Adapter.js
// Change the url as required

var server_ip = "0.0.0.0";
var port = "2000";
var url = "http://"+server_ip+":"+port+"/";

function is_defined(x) {
    return typeof x !== 'undefined';
}

function is_object(x) {
    return Object.prototype.toString.call(x) === "[object Object]";
}

function is_array(x) {
    return Object.prototype.toString.call(x) === "[object Array]";
}

var DBAdapter = {};

// CRUD: createOne.
// "collection" must be specified by the first parameter.
// "data" must be specified as an object passed in by the second parameter.
// Example,
// collection : "table1",
// data : {"entity1" : 1}
DBAdapter.createOne = function(collection, data, successCB, errorCB) {
    if (is_defined(collection) && is_defined(data) && !is_array(data)) {
        var body = {
            collection : collection,
            data : JSON.stringify(data)
        };
        ajaxCall("create", body, successCB, errorCB);
    } else {
        // Error handling
        errorCB("Error: " + "Invalid Parameters");
    }
};

// CRUD: createMany.
// "collection" must be specified in by the first parameter.
// "data" must be specified as an Array by the second parameter.
// Example,
// collection : "table1",
// data : [{"entity1" : 1}, {"entity2" : 2}]
DBAdapter.createMany = function(collection, data, successCB, errorCB) {
    if (is_defined(collection) && is_defined(data) && is_array(data)) {
        var body = {
            collection : collection,
            data : JSON.stringify(data)
        };
        ajaxCall("create", body, successCB, errorCB);
    } else {
        // Error handling
        errorCB("Error: " + "Invalid Parameters");
    }
};

// CRUD: ReadOne.
// "collection" must be specified in by the first parameter.
// "data" must be specified as an object by the second parameter.
// Example,
// collection : "table1",
// data : {"_id", "57d26068f2a81b5d740f695c"} or
// data : {"x" : 1234}
DBAdapter.readOne = function(collection, data, successCB, errorCB) {
    if (is_defined(collection) && is_defined(data)) {
        var body;
        if (is_defined(data._id)) {
            body = {
                collection : collection,
                _id : data._id
            };
        } else {
            body = {
                collection : collection,
                data : JSON.stringify(data)
            };
        }
        ajaxCall("readOne", body, successCB, errorCB);
    } else {
        // Error handling
        errorCB("Error: " + "Invalid Parameters");
    }
};

// CRUD: readAll.
// "collection" must be specified in by the first parameter.
// "data" must be specified as an object by the second parameter.
// Example,
// collection : "table1",
// data : {"x" : 1234}
DBAdapter.readAll = function(collection, body, successCB, errorCB) {
    if (is_defined(collection) && is_defined(data)) {
        var body = {
            collection : collection,
            data : JSON.stringify(data)
        };
        ajaxCall("readAll", body, successCB, errorCB);
    } else {
        // Error handling
        errorCB("Error: " + "Invalid Parameters");
    }
};

// CRUD: update.
// "collection" must be specified in by the first parameter.
// "data" must be specified as an object by the second parameter.
// Example,
// collection : "table1",
// data : {_id : "57d25c9cf2a81b5d740f6956", newData : {x:5678, y:2222}} or
// data : {oldData: {y:2222}, newData : {z:5678, y:2222}}
DBAdapter.update = function(collection, data, successCB, errorCB) {
    if (is_defined(collection) && is_defined(data)) {
        var body;
        if (is_defined(data._id) && is_defined(data.newData) && !is_array(data.newData)) {
            body = {
                collection : collection,
                _id : data._id,
                newData: JSON.stringify(data.newData)
            };
        } else if (is_defined(data.oldData) && is_defined(data.newData) &&
            !is_array(data.oldData) && !is_array(data.newData)) {
            body = {
                collection : collection,
                oldData : JSON.stringify(data.oldData),
                newData: JSON.stringify(data.newData)
            };
        } else {
            errorCB("Error: " + "Invalid Parameters");
            return;
        }
        ajaxCall("update", body, successCB, errorCB);
    } else {
        // Error handling
        errorCB("Error: " + "Invalid Parameters");
    }
};

// CRUD: delete.
// "collection" must be specified in by the first parameter.
// "data" must be specified as an object by the second parameter.
// Example,
// collection : "table1",
// data : {"x" : [{"x":1111}, {"y":3333}]} or
// data : {"y" : 2222} or
// data : {"_id" : "57d26068f2a81b5d740f695c"}
DBAdapter.delete = function(collection, data, successCB, errorCB) {
    if (is_defined(collection) && is_defined(data)) {
        var body;
        if (is_defined(data._id)) {
            body = {
                collection : collection,
                _id : data._id
            };
        } else {
            body = {
                collection : collection,
                data : JSON.stringify(data)
            };
        }
        ajaxCall("delete", body, successCB, errorCB);
    } else {
        // Error handling
        errorCB("Error: " + "Invalid Parameters");
    }
};

function ajaxCall(operation, body, successCB, errorCB) {
    $.ajax({
        "url": url + operation,
        "method": "POST",
        "content-Type": "application/json; charset=utf-8",
        "data": body,
        "success": function(result){
            successCB(result);
        },
        "error":  function(xhr, ajaxOptions, thrownError) {
            errorCB("Error: " + thrownError);
        }
    });
}





